### PR TITLE
Provide clients with additional information on failures

### PR DESCRIPTION
**Pull request by dmage**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1578269.
<details>
<summary>Output of `docker pull 172.30.1.1:5000/myproject/foo` before patch</summary>

> Using default tag: latest
> Error response from daemon: error parsing HTTP 404 response body: json: cannot unmarshal number 727750330070774258508461169329606316441540384616086620444793670799490968006862819122321718201191797419997997401183211852380009717481319078856159647052207087382780934435083321402232281992602221429720237551009157906242822670434372156158838809966012277928394724347562504363760164573515974160075643611066621396686345711634830402036967312043372357208615635121399023173317333881066377186081786347073957563537028929133941922400310822439373446952424805314987188980795445055652506350160178856351642641065671487458569247708945840257278206730396350234500284367692556456257479250642766444519821456090445111671154640466902500856558735083798101314364740720353254006753374529893599705840886037424384346242284644187703740362832697533513090048750490375674016760034300072086028322217710792288813653844592714999239800758061025142993975932729652516956614486996960592581777085180943551085591808753836186109764730481744370341292528223540427172722671373576153438610225121089005098139559328055784644568500270056805864973207750467184608923119528645893414817914609461906708700925267688487587984898849951818696290565686738016475014897465637325294372182730686094785654892785249695064105173780307688591668277268235731259066257966161189978847409961994604808674697 into Go struct field Error.detail of type float64: "{\"errors\":[{\"code\":\"MANIFEST_UNKNOWN\",\"message\":\"manifest unknown\",\"detail\":{\"Op\":\"Get\",\"URL\":\"https://10.34.129.131:5000/v2/\",\"Err\":{\"Certificate\":{\"Raw\":\"MIIE7TCCAtWgAwIBAgIJALyR26vAX3/RMA0GCSqGSIb3DQEBCwUAMBYxFDASBgNVBAMMC2V4YW1wbGUuY29tMB4XDTE4MDUyMjEyMjIxNVoXDTE5MDUyMjEyMjIxNVowFjEUMBIGA1UEAwwLZXhhbXBsZS5jb20wggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQCyYr/r/3W1hbmjMHMFQ3ETGNEQ9c+TYUDb7Min7+7zIzvrsJjgrJpMYvTD/ELPfpOzPPL2vCIKOdnARhl4H0G4yzw4nz0jMZkWrFD8b1qTUZq389CsK6Gstj65JPiPe0y/re75KI+zbhnif1wXcwJKvFwjLSlK0wGk9ejhDDt1E5EJM4M49ihTp7KE9XDb79z9WXkH1YtXpD0lwgnIryTlRo+HSOBkKZSAguY1CQ+tL/S+v8/t7XvBHWBEJuGgxdpzngkr8qzM39IaGyBpPyPEFi4+xOyigzZd9xY96nKELQD8HT1AFWuhD1Y59R/Vb6irP1GTDWK0leSUJy32qNug2QO2lC0bsZZ1kV8p650TgHF7VGEHP8SNNQucqfStpdLbfBS57QM2ShgQ22p4hzgtMZUqKhaHB0/MZEZv0WGuxtZulME/cqizayGU28zHBkkW6msnRP1DPFB5Xdksqpz07xgbg2Reqetl23fd/ZYjuZBf+0u8ejqcn2jtzllkHzhxMGWEDpKgd2GINY5360WiB60aFsTW/EVhI5NzIxmBfaUrrQaHTX3IdCcm7txf9Ms4UGcERcWgS5CgdjqoI60agZKcjZCWANOytSvEJyMXC66Uqb67APlDuzy55w4zP4hWklF3TmDz1v3w4E8vYSvgh3Kv57C5HPcQ6siXo5qdiQIDAQABoz4wPDALBgNVHQ8EBAMCBaAwLQYDVR0RBCYwJIILZXhhbXBsZS5jb22CD3d3dy5leGFtcGxlLmNvbYcEwAACATANBgkqhkiG9w0BAQsFAAOCAgEAdWWNGafabuVsTc5Bvgh7KOx4NkUGtYXVRF109eZzL72OfOvJia9qp7bfoTNgATDjxbXCKJ5kI35rt9d8Z71i1e41KSWAMiogEI9RSQ6T6LbusuYMZynIbpnrCxsYaENEYG0PCHNysB9uFotDuNImKMD+6uYX/RmVPb5uk82Wb56wevxpP3tr6062/AjZ5L0rFKAYwlsaCUM2nfv648hct0DHwtWd8LWU44QT076F4jsNiQXE21mcgWrZmplSon1RH/RFruZmUkwuuikHMKjYRbOqfYyIwSpga2ihkFOHIzPs9JeYvoncDfPC+wUS0E5M9yb2LKtPBQad2/ugx9ShMVE31MB234zRVHFxkimSp4LoYXCoGpqX7HvxmkYQVl1BGFjumuR7EiHUgWOqIzd4tLrzvDpN6Ez69Jp1BjSgDQrE9TB70npqkyx8aGTvJfQ6kDxOULE4LiUBUCQWj+jxd4EFCbDNK+CaHExbdrh9CJ8eefUANK1+JXQTkSnbdSw+PUfE87nRRZyq29/BfQRbeZ95/HovJNcy+cyUs9BlGkhrVHfMnoTarMZhaN2hqS0Wjr/f6WiTf8iXAD0S5Y0mxcSzuFDKxZMTsqt04hcmiiWHcZUwNoHMmxaPhV7F4HcQQq2JAQHZxryAn/u+3AltdonBIQn12d6G5cXWjVzeSAk=\",\"RawTBSCertificate\":\"MIIC1aADAgECAgkAvJHbq8Bff9EwDQYJKoZIhvcNAQELBQAwFjEUMBIGA1UEAwwLZXhhbXBsZS5jb20wHhcNMTgwNTIyMTIyMjE1WhcNMTkwNTIyMTIyMjE1WjAWMRQwEgYDVQQDDAtleGFtcGxlLmNvbTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBALJiv+v/dbWFuaMwcwVDcRMY0RD1z5NhQNvsyKfv7vMjO+uwmOCsmkxi9MP8Qs9+k7M88va8Igo52cBGGXgfQbjLPDifPSMxmRasUPxvWpNRmrfz0Kwroay2Prkk+I97TL+t7vkoj7NuGeJ/XBdzAkq8XCMtKUrTAaT16OEMO3UTkQkzgzj2KFOnsoT1cNvv3P1ZeQfVi1ekPSXCCcivJOVGj4dI4GQplICC5jUJD60v9L6/z+3te8EdYEQm4aDF2nOeCSvyrMzf0hobIGk/I8QWLj7E7KKDNl33Fj3qcoQtAPwdPUAVa6EPVjn1H9VvqKs/UZMNYrSV5JQnLfao26DZA7aULRuxlnWRXynrnROAcXtUYQc/xI01C5yp9K2l0tt8FLntAzZKGBDbaniHOC0xlSoqFocHT8xkRm/RYa7G1m6UwT9yqLNrIZTbzMcGSRbqaydE/UM8UHld2SyqnPTvGBuDZF6p62Xbd939liO5kF/7S7x6OpyfaO3OWWQfOHEwZYQOkqB3YYg1jnfrRaIHrRoWxNb8RWEjk3MjGYF9pSutBodNfch0Jybu3F/0yzhQZwRFxaBLkKB2OqgjrRqBkpyNkJYA07K1K8QnIxcLrpSpvrsA+UO7PLnnDjM/iFaSUXdOYPPW/fDgTy9hK+CHcq/nsLkc9xDqyJejmp2JAgMBAAGjPjA8MAsGA1UdDwQEAwIFoDAtBgNVHREEJjAkggtleGFtcGxlLmNvbYIPd3d3LmV4YW1wbGUuY29thwTAAAIB\",\"RawSubjectPublicKeyInfo\":\"MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAsmK/6/91tYW5ozBzBUNxExjREPXPk2FA2+zIp+/u8yM767CY4KyaTGL0w/xCz36Tszzy9rwiCjnZwEYZeB9BuMs8OJ89IzGZFqxQ/G9ak1Gat/PQrCuhrLY+uST4j3tMv63u+SiPs24Z4n9cF3MCSrxcIy0pStMBpPXo4Qw7dRORCTODOPYoU6eyhPVw2+/c/Vl5B9WLV6Q9JcIJyK8k5UaPh0jgZCmUgILmNQkPrS/0vr/P7e17wR1gRCbhoMXac54JK/KszN/SGhsgaT8jxBYuPsTsooM2XfcWPepyhC0A/B09QBVroQ9WOfUf1W+oqz9Rkw1itJXklCct9qjboNkDtpQtG7GWdZFfKeudE4Bxe1RhBz/EjTULnKn0raXS23wUue0DNkoYENtqeIc4LTGVKioWhwdPzGRGb9FhrsbWbpTBP3Kos2shlNvMxwZJFuprJ0T9QzxQeV3ZLKqc9O8YG4NkXqnrZdt33f2WI7mQX/tLvHo6nJ9o7c5ZZB84cTBlhA6SoHdhiDWOd+tFogetGhbE1vxFYSOTcyMZgX2lK60Gh019yHQnJu7cX/TLOFBnBEXFoEuQoHY6qCOtGoGSnI2QlgDTsrUrxCcjFwuulKm+uwD5Q7s8uecOMz+IVpJRd05g89b98OBPL2Er4Idyr+ewuRz3EOrIl6OanYkCAwEAAQ==\",\"RawSubject\":\"MBYxFDASBgNVBAMMC2V4YW1wbGUuY29t\",\"RawIssuer\":\"MBYxFDASBgNVBAMMC2V4YW1wbGUuY29t\",\"Signature\":\"dWWNGafabuVsTc5Bvgh7KOx4NkUGtYXVRF109eZzL72OfOvJia9qp7bfoTNgATDjxbXCKJ5kI35rt9d8Z71i1e41KSWAMiogEI9RSQ6T6LbusuYMZynIbpnrCxsYaENEYG0PCHNysB9uFotDuNImKMD+6uYX/RmVPb5uk82Wb56wevxpP3tr6062/AjZ5L0rFKAYwlsaCUM2nfv648hct0DHwtWd8LWU44QT076F4jsNiQXE21mcgWrZmplSon1RH/RFruZmUkwuuikHMKjYRbOqfYyIwSpga2ihkFOHIzPs9JeYvoncDfPC+wUS0E5M9yb2LKtPBQad2/ugx9ShMVE31MB234zRVHFxkimSp4LoYXCoGpqX7HvxmkYQVl1BGFjumuR7EiHUgWOqIzd4tLrzvDpN6Ez69Jp1BjSgDQrE9TB70npqkyx8aGTvJfQ6kDxOULE4LiUBUCQWj+jxd4EFCbDNK+CaHExbdrh9CJ8eefUANK1+JXQTkSnbdSw+PUfE87nRRZyq29/BfQRbeZ95/HovJNcy+cyUs9BlGkhrVHfMnoTarMZhaN2hqS0Wjr/f6WiTf8iXAD0S5Y0mxcSzuFDKxZMTsqt04hcmiiWHcZUwNoHMmxaPhV7F4HcQQq2JAQHZxryAn/u+3AltdonBIQn12d6G5cXWjVzeSAk=\",\"SignatureAlgorithm\":4,\"PublicKeyAlgorithm\":1,\"PublicKey\":{\"N\":727750330070774258508461169329606316441540384616086620444793670799490968006862819122321718201191797419997997401183211852380009717481319078856159647052207087382780934435083321402232281992602221429720237551009157906242822670434372156158838809966012277928394724347562504363760164573515974160075643611066621396686345711634830402036967312043372357208615635121399023173317333881066377186081786347073957563537028929133941922400310822439373446952424805314987188980795445055652506350160178856351642641065671487458569247708945840257278206730396350234500284367692556456257479250642766444519821456090445111671154640466902500856558735083798101314364740720353254006753374529893599705840886037424384346242284644187703740362832697533513090048750490375674016760034300072086028322217710792288813653844592714999239800758061025142993975932729652516956614486996960592581777085180943551085591808753836186109764730481744370341292528223540427172722671373576153438610225121089005098139559328055784644568500270056805864973207750467184608923119528645893414817914609461906708700925267688487587984898849951818696290565686738016475014897465637325294372182730686094785654892785249695064105173780307688591668277268235731259066257966161189978847409961994604808674697,\"E\":65537},\"Version\":3,\"SerialNumber\":13587883081466871761,\"Issuer\":{\"Country\":null,\"Organization\":null,\"OrganizationalUnit\":null,\"Locality\":null,\"Province\":null,\"StreetAddress\":null,\"PostalCode\":null,\"SerialNumber\":\"\",\"CommonName\":\"example.com\",\"Names\":[{\"Type\":[2,5,4,3],\"Value\":\"example.com\"}],\"ExtraNames\":null},\"Subject\":{\"Country\":null,\"Organization\":null,\"OrganizationalUnit\":null,\"Locality\":null,\"Province\":null,\"StreetAddress\":null,\"PostalCode\":null,\"SerialNumber\":\"\",\"CommonName\":\"example.com\",\"Names\":[{\"Type\":[2,5,4,3],\"Value\":\"example.com\"}],\"ExtraNames\":null},\"NotBefore\":\"2018-05-22T12:22:15Z\",\"NotAfter\":\"2019-05-22T12:22:15Z\",\"KeyUsage\":5,\"Extensions\":[{\"Id\":[2,5,29,15],\"Critical\":false,\"Value\":\"AwIFoA==\"},{\"Id\":[2,5,29,17],\"Critical\":false,\"Value\":\"MCSCC2V4YW1wbGUuY29tgg93d3cuZXhhbXBsZS5jb22HBMAAAgE=\"}],\"ExtraExtensions\":null,\"UnhandledCriticalExtensions\":null,\"ExtKeyUsage\":null,\"UnknownExtKeyUsage\":null,\"BasicConstraintsValid\":false,\"IsCA\":false,\"MaxPathLen\":0,\"MaxPathLenZero\":false,\"SubjectKeyId\":null,\"AuthorityKeyId\":null,\"OCSPServer\":null,\"IssuingCertificateURL\":null,\"DNSNames\":[\"example.com\",\"www.example.com\"],\"EmailAddresses\":null,\"IPAddresses\":[\"192.0.2.1\"],\"URIs\":null,\"PermittedDNSDomainsCritical\":false,\"PermittedDNSDomains\":null,\"ExcludedDNSDomains\":null,\"PermittedIPRanges\":null,\"ExcludedIPRanges\":null,\"PermittedEmailAddresses\":null,\"ExcludedEmailAddresses\":null,\"PermittedURIDomains\":null,\"ExcludedURIDomains\":null,\"CRLDistributionPoints\":null,\"PolicyIdentifiers\":null},\"Host\":\"10.34.129.131\"}}}]}\n"
</details>

### CODE REVIEW

**Comment by dmage** [comment]
@miminar @bparees @legionus do you see any better solutions? The returned error will be printed to the log and sent to the client in the Detail field as JSON.

The problematic error is a struct with an exportable BigInt field. The client can't parse large integers. Though, this field isn't used by the client at all.

This fix doesn't change the message in the log, but Detail will be just `{}`. But it fixes this particular bug, not the problem itself.

**Comment by dmage** [comment]
@smarterclayton do we need to run tests for [Feature:DeploymentConfig] in this repo?

**Comment by bparees** [comment]
@dmage so it will get rid of this text: `cannot unmarshal number AB2344... into Go struct field Error.detail of type float64:` but the rest of the output (as seen in the BZ) will be the same?

or it will be:
```
"{\"errors\":[{\"code\":\"MANIFEST_UNKNOWN\",\"message\":\"manifest unknown\",\"detail\" : {}
```

?

because that seems worse (we've lost any indication there's a certificate issue)

**Comment by bparees** [comment]
I guess there is no way for us to inspect the error we got back and understand it was a certificate issue?  it bothers me that the error doesn't really give any hint about what is really wrong here.

**Comment by miminar** [comment]
> This fix doesn't change the message in the log, but Detail will be just {}.

Can the `Detail` be replaced by something parse-able?

**Comment by dmage** [comment]
@bparees we need to patch the upstream code to add any information for the client.

https://github.com/openshift/docker-distribution/blob/4e8e323b4b83451e20a564bc40e3b26968b3a887/registry/handlers/images.go#L96-L100

They unconditionally wraps our error with ErrorCodeManifestUnknown, so we'll need to patch this. But after that we'll face another problem: their error types doesn't provide mechanisms to change the error message, so we'll need to bypass their error type system.

@miminar we can provide anything we want for this field, the question is what we want it to be. Because the client ignores this field, it's unclear to me what it should be.

**Comment by miminar** [comment]
> Because the client ignores this field, it's unclear to me what it should be.

What if we patch docker to print the details with debug loglevel? It makes sense not only for this specific use case.

**Comment by dmage** [comment]
I've updated my PR, now it breaks the upstream type system. Though, it solves the problem.

On the client now I can see the error:
```
$ docker pull 172.30.1.1:5000/myproject/foo
Using default tag: latest
Error response from daemon: unknown: Get https://10.34.129.131:5000/v2/: x509: certificate is valid for 192.0.2.1, not 10.34.129.131
```

@legionus @miminar PTAL

**Comment by legionus** [comment]
looks good.

File: pkg/dockerregistry/server/pullthroughmanifestservice.go [hunk 1]
```diff
@@ -66,7 +67,7 @@ func (m *pullthroughManifestService) remoteGet(ctx context.Context, dgst digest.
 	repo, err := m.getRemoteRepositoryClient(ctx, &ref, dgst, options...)
 	if err != nil {
 		context.GetLogger(ctx).Errorf("error getting remote repository for image %q: %v", ref.Exact(), err)
-		return nil, err
+		return nil, errors.ErrorCodeFIXME.WithArgs(err)
```
Comment (miminar) [inline pkg/dockerregistry/server/pullthroughmanifestservice.go]: ManifestUnknown should probably passed as is.

File: pkg/errors/errors.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"net/http"
+
+	errcode "github.com/docker/distribution/registry/api/errcode"
+)
+
+const errGroup = "openshift"
+
+var (
+	ErrorCodeFIXME = errcode.Register(errGroup, errcode.ErrorDescriptor{
```
Comment (miminar) [inline pkg/errors/errors.go]: Having a dedicated error for certificate issues would be nice. Telling the certificate error from any other may not be that easy though.

Unless we know of any other unparse-able error, we could leave the rest of them unchanged and come back later when someone discovers one.

:+1: for this file. I bet there's gonna be more OpenShift specific errors.

Comment (bparees) [reply pkg/errors/errors.go]: any objections to renaming it "ErrorCodeWrapper" ?

File: vendor/github.com/docker/distribution/registry/handlers/images.go [hunk 3]
```diff
@@ -95,7 +95,13 @@ func (imh *imageManifestHandler) GetImageManifest(w http.ResponseWriter, r *http
...
 	manifest, err = manifests.Get(imh, imh.Digest, options...)
 	if err != nil {
-		imh.Errors = append(imh.Errors, v2.ErrorCodeManifestUnknown.WithDetail(err))
+		switch err.(type) {
+		case errcode.ErrorCode, errcode.Error:
+			// keep known error as is
+		default:
+			err = v2.ErrorCodeManifestUnknown.WithDetail(err)
+		}
+		imh.Errors = append(imh.Errors, err)
```
Comment (bparees) [inline vendor/github.com/docker/distribution/registry/handlers/images.go]: we need to do this as a carry commit to our fork of the docker distribution repo and then bump it here, right?

Comment (bparees) [reply vendor/github.com/docker/distribution/registry/handlers/images.go]: (and probably add a test to make sure we don't accidentally drop the commit at some point)

**Pull request merged**